sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,310 @@
+            {
+                //HttpClientHandler.ServerCertificateCustomValidationCallback uses Func<HttpRequestMessage, X509Certificate2, X509Chain, SslPolicyErrors, bool>
+                //We're actually looking for Func<Any Sender, X509Certificate2, X509Chain, SslPolicyErrors, bool>
+                var parameters = NamedSymbol.DelegateInvokeMethod.Parameters;
+                return parameters.Length == 4   //And it should! T1, T2, T3, T4
+                    && parameters[0].Type.IsClassOrStruct() //We don't care about common (Object) nor specific (HttpRequestMessage) type of Sender
+                    && parameters[1].Type.Is(KnownType.System_Security_Cryptography_X509Certificates_X509Certificate2)
+                    && parameters[2].Type.Is(KnownType.System_Security_Cryptography_X509Certificates_X509Chain)
+                    && parameters[3].Type.Is(KnownType.System_Net_Security_SslPolicyErrors)
+                    && NamedSymbol.DelegateInvokeMethod.ReturnType.Is(KnownType.System_Boolean);
+            }
+            return false;
+        }
+
+        private static ImmutableArray<Location> ArgumentLocations(InspectionContext c, ExpressionSyntax expression)
+        {
+            var ret = ImmutableArray.CreateBuilder<Location>();
+            ISymbol MS;
+            switch (expression)
+            {
+                case IdentifierNameSyntax identifier:
+                    MS = c.Context.SemanticModel.GetSymbolInfo(identifier).Symbol;
+                    if (MS != null)
+                    {
+                        foreach (var syntax in MS.DeclaringSyntaxReferences.Select(x => x.GetSyntax()))
+                        {
+                            ret.AddRange(IdentifierLocations(c, syntax));
+                        }
+                    }
+                    break;
+                case ParenthesizedLambdaExpressionSyntax lambda:
+                    if ((lambda.Body as LiteralExpressionSyntax)?.Kind() == SyntaxKind.TrueLiteralExpression)
+                    {
+                        ret.Add(lambda.Body.GetLocation());   //Code was found guilty for lambda (...) => true
+                    }
+                    else if (lambda.Body is BlockSyntax block)
+                    {
+                        ret.AddRange(BlockLocations(block));
+                    }
+                    break;
+                case InvocationExpressionSyntax invocation:
+                    MS = c.Context.SemanticModel.GetSymbolInfo(invocation).Symbol;
+                    if (MS != null)
+                    {
+                        foreach (var syntax in MS.DeclaringSyntaxReferences.Select(x => x.GetSyntax() as MethodDeclarationSyntax).Where(x => x != null))
+                        {
+                            c.VisitedMethods.Add(syntax);
+                            ret.AddRange(InvocationLocations(c, syntax));
+                        }
+                    }
```

Comment:
Same question/remark as on line 124: if I replace the for-loop, I don't see any change in unit tests:
```
                            if (MS != null && MS.DeclaringSyntaxReferences.Length == 1 && MS.DeclaringSyntaxReferences[0].GetSyntax() is MethodDeclarationSyntax syntax)
                            {
                                c.VisitedMethods.Add(syntax);
                                ret.AddRange(InvocationLocations(c, syntax));
                            }
```
So I wonder if the `foreach` is truly necessary, as I don't believe we can create a case where the declaration syntax has several nodes.